source/bdDetect.py

```diff
@@ -0,0 +1,427 @@
+	"""Get any matching drivers for possible Bluetooth devices.
+	@return: Pairs of drivers and port information.
+	@rtype: generator of (str, L{DeviceMatch}) tuples
+	"""
+	btDevs = [DeviceMatch(KEY_SERIAL, port["bluetoothName"], port["port"], port)
+		for port in hwPortUtils.listComPorts()
+		if "bluetoothName" in port]
+	for match in btDevs:
+		for driver, devs in _driverDevices.iteritems():
+			matchFunc = devs[KEY_BLUETOOTH]
+			if not callable(matchFunc):
+				continue
+			if matchFunc(match):
+				yield driver, match
+
+class Detector(object):
+	"""Automatically detect braille displays.
+	This should only be used by the L{braille} module.
+	"""
+
+	def __init__(self):
+		self._BgScanApc = winKernel.PAPCFUNC(self._bgScan)
+		self._btComs = None
+		self._detectUsb = False
+		self._detectBluetooth = False
+		core.windowMessageReceived.register(self.handleWindowMessage)
+		self._stopEvent = threading.Event()
+		self._queuedSscanLock = threading.Lock()
+		self._scanQueued = False
+		# Perform initial scan.
+		self._startBgScan(usb=True, bluetooth=True)
+
+	def _startBgScan(self, usb=False, bluetooth=False):
+		self._detectUsb = usb
+		self._detectBluetooth = bluetooth
+		with self._queuedSscanLock:
+			if not self._scanQueued:
+				braille._BgThread.queueApc(self._BgScanApc)
+				self._scanQueued = True
+
+	def _stopBgScan(self):
+		self._stopEvent.set()
+
+	def _bgScan(self, param):
+		self._stopEvent.clear()
+		with self._queuedSscanLock:
+			self._scanQueued = False
+		if self._detectUsb:
+			if self._stopEvent.isSet():
+				return
+			for driver, match in getDriversForConnectedUsbDevices():
+				if self._stopEvent.isSet():
+					return
+				if braille.handler.setDisplayByName(driver, detected=match):
+					return
+		if self._detectBluetooth:
+			if self._stopEvent.isSet():
+				return
+			if self._btComs is None:
+				btComs = list(getDriversForPossibleBluetoothDevices())
+				# Cache Bluetooth com ports for next time.
+				btComsCache = []
+			else:
+				btComs = self._btComs
+				btComsCache = btComs
+			for driver, match in btComs:
+				if self._stopEvent.isSet():
+					return
+				if btComsCache is not btComs:
+					btComsCache.append((driver, match))
+				if braille.handler.setDisplayByName(driver, detected=match):
+					return
+			if self._stopEvent.isSet():
+				return
+			if btComsCache is not btComs:
+				self._btComs = btComsCache
+
+	def rescan(self):
+		"""Stop a current scan when in progress, and start scanning from scratch."""
+		self._stopBgScan()
+		# A Bluetooth com port might have been added.
+		self._btComs = None
+		self._startBgScan(usb=True, bluetooth=True)
+
+	def handleWindowMessage(self, msg=None, wParam=None):
+		if msg == WM_DEVICECHANGE and wParam == DBT_DEVNODES_CHANGED:
+			self.rescan()
+
+	def pollBluetoothDevices(self):
+		"""Poll bluetooth devices that might be in range.
+		This does not cancel the current scan and only queues a new scan when no scan is in progress."""
+		if not self._btComs:

```

Comment:
Is this ran in the same thread as the scanning function? If not, `_btComs` may be in the middle of being updated.